web/src/stream_data.js

```diff
@@ -368,10 +369,15 @@ export function get_subscribed_streams_for_user(user_id) {
         }
         if (is_user_subscribed(sub.stream_id, user_id)) {
             subscribed_subs.push(sub);
+        } else {
+            can_subscribe_subs.push(sub);
         }
     }
 
-    return subscribed_subs;
+    return {
+        subscribed: subscribed_subs,
+        can_subscribe: can_subscribe_subs,
```

Comment:
This feels a bit misleading of a name -- we haven't verified that the user can subscribe themselves to the stream -- just that they can see the stream's subscribers. It may happen to be equivalent in many cases, I guess?

Reply:
So, I reviewed some cases. If the user does not have permission to subscribe themselves to a certain stream, then that private stream doesn't appear in the can_subscribe_subs. In the end, the user can actually subscribe to all streams. However, if you believe there might be a more appropriate name for this, please let me know; I'll make the change accordingly.

Reply:
I think the bug is you should be checking `else if (can_subscribe_others(sub)`.

Reply:
```diff --git a/web/src/stream_data.js b/web/src/stream_data.js
index f11d682b02..e91bbe2e1b 100644
--- a/web/src/stream_data.js
+++ b/web/src/stream_data.js
@@ -369,7 +369,7 @@ export function get_streams_for_user(user_id) {
         }
         if (is_user_subscribed(sub.stream_id, user_id)) {
             subscribed_subs.push(sub);
-        } else {
+        } else if (can_subscribe_others(sub)) {
             can_subscribe_subs.push(sub);
         }
     }
diff --git a/web/tests/stream_data.test.js b/web/tests/stream_data.test.js
index 5b00cda434..ecb69203ac 100644
--- a/web/tests/stream_data.test.js
+++ b/web/tests/stream_data.test.js
@@ -256,7 +256,12 @@ test("get_streams_for_user", () => {
         denmark,
         social,
     ]);
+    assert.deepEqual(stream_data.get_streams_for_user(test_user.user_id).can_subscribe, []);
+
+    // Verify can subscribe if we're an administrator.
+    page_params.is_admin = true;
     assert.deepEqual(stream_data.get_streams_for_user(test_user.user_id).can_subscribe, [world]);
+    page_params.is_admin = false;
 });
 
 test("renames", () => {
```

Reply:
Thanks for the suggestion, I have made the suggested changes.